### ISSUE TITLE
Address potential burden on project maintainers

### ISSUE THREAD

**Issue by ncoghlan** [pullrequestday/pullrequestday#21]
While the idea of having a global "start contributing to open source day" is an interesting one, it's also entirely possible for such events to backfire and create more work for already overloaded volunteer maintainers.

Ian Cordasco discusses some of those potential problems in his post at http://www.coglib.com/~icordasc/blog/2015/11/corporations-and-oss-do-not-mix.html

The most relevant problem is that for quite a few open source projects, the problems for maintainers aren't whether or not submissions are coming in, but rather the quality of those submissions, and the time needed to review them.

OpenHatch aims to connect projects looking for new contributors with folks looking for projects to contribute to, so it's potentially worth talking to them about possible ways to address this concern while still preserving the "Get involved upstream!" spirit of the event.

(If I've correctly understood that intended spirit, then a name like "Swim Upstream Day" seems to better reflect that intent than simply "Pull Request Day", as the latter suggests one-off contributions to me, rather than an ongoing change to the way we interact with a community)

**Comment by rossjones** [comment]
Would #22 and #23 solve the possible problems?  I think being opt-in, and providing a way for smaller projects (of which many are under-resourced) to mentor potential contributors might go some way to alleviating you fears?

**Comment by sigmavirus24** [comment]
So what I think would be far more valuable (and obviously an order of magnitude more difficult) would be to start to reach out to projects (and recruit projects) to do the following:
1. Provide clear, concise, correct, and complete documentation for getting started with developing a project
2. Create one label that is uniform across all participating projects (e.g., 'Beginner Friendly', 'Low Hanging Fruit', 'Easy', etc.) without trying to find consensus and just picking one (because consensus building will waste too much time)
3. Encourage, remind, and help project owners/maintainers continue to label issues that way
4. Load balance the contributors we send to those projects, e.g., if there are 100 Ruby projects, when we suggest one to a new participant, select projects to display based on how many people have seen that project/clicked through to possibly contribute.

This means that as you start aggregating projects that wish to participate (either by them volunteering or through you recruiting them) they are prepared to set new contributors up for success. At the same time, contributors should be actively recruited as well and they should be encouraged to start looking at issues with the designated tag. They should start discussing the issue with the maintainers, reporter, etc. and they should start learning the code base as soon as possible. No project will have the resources to mentor all potential contributors on that day. Also, some projects (e.g., pretty much any project I operate) offers mentoring to potential contributors all year round. However, I can not mentor more than a single person a day. As an introvert, it's draining for me to mentor one person I don't know in a single day, let alone 2 or 3 or 5.

If we want this event to be full of success stories (I made my first contribution to an open source project!) then the emphasis needs to not be on "We need to submit 1 million pull requests on this exact day (also timezones will be fun for you. Does your day stretch from 000000 UTC+12 to 235959 UTC-12?, etc.). The emphasis should be on "Each of these people has promised to send at least 1 pull request on this exact day, collaborated with the project and sent an awesome PR that has been accepted."

Further, I would suggest a cut-off day for participants on both sides (projects and contributors). Projects should register with the event 2 months in advance and as soon as we're within the two month range of the event, no new projects should be added. Some projects may be able to prepare the above requirements, but adding new projects so close to the day of the event makes it hard for contributors to explore all of their options thoroughly.

Likewise, new contributor registration should stop at some point too. If someone signs up the day before, they will likely not easily find a project and issue they want to work on or a mentor to help. This will be frustrating for everyone involved.

**Comment by sigmavirus24** [comment]
(Sorry for commenting so shortly after my last, but this just came to me)

I would also argue that any project that's willing to encourage and mentor new contributors would be willing to add a Code of Conduct to their project. I would further argue, that to encourage a truly diverse group of new contributors, that this event should strongly suggest (if not require) Codes of Conduct be in place before accepting a project into the event. This event should even compile a list of Codes of Conduct that it endorses. I would argue it should disallow the NCoC (No Code of Conduct) and the ineffective CoC that LKML adopted for Linux development.

If this event really is meant for new contributors, let's ensure that the new contributors feel safe (not just empowered) when contributing to these projects.

**Comment by nedbat** [comment]
I think it would also be helpful to provide some guidance about what makes a useful pull request.  I know the list of requirements can seem daunting, and might discourage contributors.  But if I get a pull request without tests, that hasn't taken into account all the support requirements, and so on, then the contributor will still get discouraged, but not until after I have to tell them "no."
Can we add to the pledge something like, "I will read the contribution guidelines for the repo, and will consider all the factors suggested there."  ?

**Comment by sigmavirus24** [comment]
@nedbat yeah, that's something that was kind of implied in my line of reasoning around guides for new contributors getting started and such. Thank you for making that explicit

**Comment by M-Zuber** [comment]
With regards to projects being contribution ready, maybe it would be worth it to set up clear guidelines and have submitted projects vetted via pull requests.

**Comment by sigmavirus24** [comment]
@M-Zuber Absolutely. People pledging to send PRs in February are already added via pull request. It makes sense for there to be some kind of checklist for a project to complete before being added as well.

**Comment by M-Zuber** [comment]
That would require another part  of the site, or minimum a file that can be included into the site once the deadline for adding projects is reached.

**Comment by M-Zuber** [comment]
With regards to projects I would argue for classifications of what the maintener is looking for. 
There are projects that I deal with that are currently only looking for documentation help for example.

**Comment by sigmavirus24** [comment]
Right, so right now for [people who have pledged](http://pullrequestday.com/pledge/), there are a list of "cards" with their information. What I'm imagining is something similar for projects. Imagine something like the following

A card that links to:
- Project (just to be explicit)
- Getting started
- Issues pre-filtered with standardized label
- Code of Conduct
- People to reach out to for mentoring

If people stick to issues with the chosen label, the likelihood is that those projects will get the help they're looking for (if they've used the label appropriately).

If we want extra labels to add to cards, e.g., `Documentation`, `Code`, `Python`, `Ruby`, etc. to help people search through things that might work but it looks like the existing website is Jekyll and that might not be an attainable goal for this February (I don't know what the availability is of the organizers is to do that work or review that work).

**Comment by shaunagm** [comment]
The framing of "One million pull requests in one day" has me wary from the get-go.  What sort of goal does that imply?  If we're trying to increase the number of people contributing to FOSS, then the priority is creating a good experience for newcomers.  If we're trying to help out FOSS projects as much as possible, then the priority is creating high quality, helpful pull requests.  I think that either of those goals would be better served by [Sigmavirus'](https://github.com/pullrequestday/pullrequestday/issues/21#issuecomment-154832322) and [Ned's](https://github.com/pullrequestday/pullrequestday/issues/21#issuecomment-154833301) suggestions, and by being more organized about this.

Things that I think would help make this event more successful, in addition to what has already been mentioned:
- If possible, pairing mentors and contributors ahead of time and giving contributors information about the project to go through so they're better prepared on the day of.  The biggest single barrier to new contribution is not knowing who to ask for help when you get stuck.
- Also, someone needs to keep track of who is working on what.  You can't trust a giant mass of newcomers to navigating claiming issues with comments, so it's important to keep the maintainers involved so they can say things like, "X is already working on this issue.  Let me introduce the two of you - perhaps he would enjoy having someone to work with."
- Well-publicized channels on IRC (or maybe Slack or Twitter?) where people can hang out and ask for help from peers and from "general mentors".  Very useful for people who don't have mentors, or just to take a load off of project maintainers.  General mentors can help a great deal with "meta" questions like figuring out if you should update an issue with new information, or helping you through the development environment setup process, etc.
- Pull Request Day (PRD) organizers and other volunteers going through issues selected for PRD and giving feedback about [whether they're actually appropriate & have enough information](http://opensource-events.com/#preparing-tasks). 

<b>Edited to add:</b>  I'd also recommend partnering with organizations that focus on specific elements of FOSS contribution, like design (#opensourcedesign), documentation, translation, accessibility, etc.  Some projects want to improve in these areas but are not sure how to themselves - I imagine that people from these communities might be interested in not only helping those projects but in helping new contributors help those projects.

**Comment by M-Zuber** [comment]
@shaunagm those are some very good points.

I would recommend Gitter as a communication tool, for the following reasons:
- excellent integration with Github (which everyone will be using anyway)
- easy to set up 
- easy for others to join
- easy to create side conversations for specific projects, issues, ect. that can then have others quickly added

Do you have organizations to recommend?

**Comment by davidmiller** [comment]
Folks,

Glad to have you all here. I appreciate you taking the time out to give your feedback,
particularly given the fact that it's remarkably thoughtful, constructive feedback!

Couple of points:

For sure, "Number of Pull Requests" is not the correct metric to
optimize absent other, subtler considerations.

Any initiative that results in mass spamming of Open Source project maintainers is
not a force for good in the world. Let's not do that.

Summarising the suggestions above, a useful event would likely do the following things:
- Create a curated list of potential projects for people to contribute to.
- Actively recruit projects that want to encourage new contributors.
- Figure out _what_ projects want help with, and how to communicate with the current maintainer(s).
- Require that curated projects have a code of conduct.
- Provide some guidance about what good looks like  - there are some ways to do that
  generically, but individual projects may have specific guidelines.
- Have well publicised places where people involved in the event can assist with the burden of
  mentoring & / or reviewing contributions, particularly from less experienced folk.

Which broadly sounds super to me. (Sure there will be some details to figure out.)

The general feeling also seems to be that there's a need for people to think ahead and pick the
project they want to contribute to beforehand. I'm not sure that needs to be a hard requirement, but
it's almost certainly a good thing to encourage. That should probably include a mechanism for potential
contributors to invite projects to participate.

I'm not entirely sure how monitoring exactly who's working on what would be practical - happy to explore
if it turns out to be a tractable cat-herding problem?

Some other thoughts that have occurred to me during the day:

It strikes me that a fairly effective mechanism for reducing the potential for abuse would be changing the pledge to:

```
We, the undersigned, pledge to
make one pull request
to an open source project
on 27th February 2016.
```

...which would remove the incentive for nuisance PRs of the "Add a second full stop" variety by
"limiting" things to one per person. - Thoughts?

I'd suggest that PRs to projects that have not explicitly opted-in remain potentially useful, but perhaps we
might find a way to have them reviewed by someone involved in the event before actually
being filed against the project? (Something that more experienced people could do after submitting their one PR maybe.)

Actions from this seem to be:
- Make explicit that quality is important here, not just quantity
- Find some projects that want new contributors to seed the list
- Generally improve the clarity and depth of copy on the site to reflect this conversation as well as other broader, subtler considerations.

_if the pledge gets changed_
- Update the pledge
- Let everyone who added themselves know that we changed the text (so that they have the opportunity to opt out if they so desire).

If I've missed anyone's specific point here, apologies - it was a lot to address at once !
Please do either re-raise it here, in it's own issue, or ping me and direct me to the bit I've missed out.
